### ISSUE TITLE
Implement "Masking" of Evidence Context within Privacy Pioneer

### ISSUE THREAD

**Issue by notowen333** [privacy-tech-lab/privacy-pioneer#352]
# Background
PP finds different kinds of data: town, state, coordinates, IP, etc. In #351 we are trying to use NLP to verify regular expression matches for these data. In order to do this, we need to build a pre-processing engine within PP.

## Why
We need to be able to tokenize data contexts to
1) Build the training set for our model
2) Run the model live (takes preprocessed data as input)

## Description 
By tokenize I mean replacing specific pieces of data with generics, i.e. replace "Middletown" with "`<TARGET-CITY>` etc. This is described in more detail in #349. 

## Automation
**The big point is that we need to be able to automate this tokenization**. 

Even if we did it by hand for the training set, we pass preprocessed tokenized data to our model, so the automation needs to be developed anyway.

## Training Set

Working this issue out should drastically simplify the process to create a training set.

## Code

This code will be inside of PP but is entirely modular. Once we are ready, it will be callable where appropriate.

**Comment by SebastianZimmeck** [comment]
> We need to be able to tokenize data contexts to
> Build the training set for our model
> Run the model live (takes preprocessed data as input)

> Once we are ready, it will be callable where appropriate.

Just to clarify, we develop and run the training set (including pre-processing) outside of our PP code, right? What we do is to include the trained model.

(cc'ing @harkous)

**Comment by notowen333** [comment]
> Just to clarify, we develop and run the training set (including pre-processing) outside of our PP code, right? What we do is to include the trained model.

My understanding is that we need pre-processing for both training and execution so we might as well collect the training set within PP directly.

**Comment by SebastianZimmeck** [comment]
> My understanding is that we need pre-processing for both training and execution so we might as well collect the training set within PP directly.

Absolutely!

**Comment by notowen333** [comment]
In the above commits I have a working version of live tokenization. 

## Example output:
![Screen Shot 2021-12-12 at 10 57 42 PM](https://user-images.githubusercontent.com/51685858/145752348-604e3eff-370e-47e3-b127-e3e651ce47cd.png)

## Tokenization Question
@harkous @SebastianZimmeck 

As you can see in the image, only one token is created per context: the token for the specific piece of "evidence" we are trying to validate. So even if other user-specific data happens to occur in the context, it does not get a token. 

**Is this the correct way to approach pre-processing when using the trained model or would all pieces of user data in the context need tokens?**

**Comment by notowen333** [comment]
As discussed, the answer to this question:

> Is this the correct way to approach pre-processing when using the trained model or would all pieces of user data in the context need tokens?

Is that all data should be masked. Run-time inference should get the same shaped-data that we train the model on. 

I will work on making this change within the PP structure (i.e. passing a given users data to the masking API).

**Comment by SebastianZimmeck** [comment]
@notowen333, I unassigned you on this issue. If there is something left to do, it would be good if you can explain to @JustinCasler or @danielgoldelman.

**Comment by notowen333** [comment]
There is a rough implementation of this [here](https://github.com/privacy-tech-lab/privacy-pioneer-data-and-scripts/blob/main/nlpCrawl/src/util.js) from the crawler. The functionality is in the `tokenize` function. 

It's using the built in `str.replace` function for the regular expressions and then a custom routine for coordinates. Maybe, the custom routine could be simplified a bit.

There is a methodological question about which bound (coarse or fine) to use for coordinates. When I did the crawl, I used the coarse bound.
